src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/DecryptorPalWindows.Decrypt.cs

```diff
@@ -0,0 +1,223 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography;
+using System.Security.Cryptography.Pkcs;
+using System.Security.Cryptography.X509Certificates;
+
+using Microsoft.Win32.SafeHandles;
+
+using static Interop.Crypt32;
+
+namespace Internal.Cryptography.Pal.Windows
+{
+    internal sealed partial class DecryptorPalWindows : DecryptorPal
+    {
+        public sealed override ContentInfo TryDecrypt(RecipientInfo recipientInfo, X509Certificate2 cert, X509Certificate2Collection originatorCerts, X509Certificate2Collection extraStore, out Exception exception)
+        {
+            Debug.Assert(recipientInfo != null);
+            Debug.Assert(cert != null);
+            Debug.Assert(originatorCerts != null);
+            Debug.Assert(extraStore != null);
+
+            CryptKeySpec keySpec;
+            exception = TryGetKeySpecForCertificate(cert, out keySpec);
```

Comment:
Throwing is in appropriate because the caller may have other recipients he wants to try to decode for.

In the original, this was just an HRESULT but (1) that felt like an inappropriate thing to return through the abstract PAL interface, and (2) an exception allows to retain richer information.

As to why it’s a return type in some and an [out] in others – that’s an inconsistency.


From: Jeremy Barton [mailto:notifications@github.com]
Sent: Wednesday, April 27, 2016 10:33 AM
To: dotnet/corefx <corefx@noreply.github.com>
Cc: Atsushi Kanamori <atsushik@microsoft.com>; Author <author@noreply.github.com>
Subject: Re: [dotnet/corefx] Add Support for EnvelopedCms (#8083)


In src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/DecryptorPalWindows.Decrypt.cs<https://github.com/dotnet/corefx/pull/8083#discussion_r61301519>:

> +

> +using static Interop.Crypt32;

> +

> +namespace Internal.Cryptography.Pal.Windows

> +{

> +    internal sealed partial class DecryptorPalWindows : DecryptorPal

> +    {

> +        public sealed override ContentInfo TryDecrypt(RecipientInfo recipientInfo, X509Certificate2 cert, X509Certificate2Collection originatorCerts, X509Certificate2Collection extraStore, out Exception exception)

> +        {

> +            Debug.Assert(recipientInfo != null);

> +            Debug.Assert(cert != null);

> +            Debug.Assert(originatorCerts != null);

> +            Debug.Assert(extraStore != null);

> +

> +            CryptKeySpec keySpec;

> +            exception = TryGetKeySpecForCertificate(cert, out keySpec);

It feels really weird that you have a Try method (return data, or return null-and-out-Exception) which calls a Try method that returns an exception or outs data.

A) Why aren't they just throwing?
B) Why the paradigm flip?

—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub<https://github.com/dotnet/corefx/pull/8083/files/e131afcf6e85303ddf92473332884d583e64241a#r61301519>